lib/bpf/network.go

```diff
@@ -0,0 +1,312 @@
+
+	"github.com/gravitational/trace"
+
+	"github.com/iovisor/gobpf/bcc"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	lostNetworkEvents = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: teleport.MetricLostNetworkEvents,
+			Help: "Number of lost network events.",
+		},
+	)
+)
+
+func init() {
+	prometheus.MustRegister(lostNetworkEvents)
+}
+
+// rawConn4Event is sent by the eBPF program that Teleport pulls off the perf
+// buffer.
+type rawConn4Event struct {
+	// CgroupID is the internal cgroupv2 ID of the event.
+	CgroupID uint64
+
+	// Version is the version of TCP (4 or 6).
+	Version uint64
+
+	// PID is the process ID.
+	PID uint32
+
+	// SrcAddr is the source IP address.
+	SrcAddr uint32
+
+	// DstAddr is the destination IP address.
+	DstAddr uint32
+
+	// DstPort is the port the connection is being made to.
+	DstPort uint16
+
+	// Command is name of the executable making the connection.
+	Command [commMax]byte
+}
+
+// rawConn6Event is sent by the eBPF program that Teleport pulls off the perf
+// buffer.
+type rawConn6Event struct {
+	// CgroupID is the internal cgroupv2 ID of the event.
+	CgroupID uint64
+
+	// Version is the version of TCP (4 or 6).
+	Version uint64
+
+	// PID is the process ID.
+	PID uint32
+
+	// SrcAddr is the source IP address.
+	SrcAddr [4]uint32
+
+	// DstAddr is the destination IP address.
+	DstAddr [4]uint32
+
+	// DstPort is the port the connection is being made to.
+	DstPort uint16
+
+	// Command is name of the executable making the connection.
+	Command [commMax]byte
+}
+
+type conn struct {
+	closeContext context.Context
+
+	// v{4,6}EventCh are the channels upon which the perf buffer places
+	// events.
+	v4EventCh <-chan []byte
+	v6EventCh <-chan []byte
+
+	// v{4,6}LostCh are the channels upon which the perf buffer places lost
+	// event count.
+	v4LostCh <-chan uint64
+	v6LostCh <-chan uint64
+
+	module   *bcc.Module
+	perfMaps []*bcc.PerfMap
+}
+
+func startConn(closeContext context.Context, pageCount int) (*conn, error) {
+	var err error
+
+	e := &conn{
+		closeContext: closeContext,
+	}
+
+	e.module = bcc.NewModule(connSource, []string{})
+	if e.module == nil {
+		return nil, trace.BadParameter("failed to load libbcc")
+	}
+
+	// Hook IPv4 connection attempts.
+	err = attachProbe(e.module, "tcp_v4_connect", "trace_connect_entry")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = attachRetProbe(e.module, "tcp_v4_connect", "trace_connect_v4_return")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Hook IPv6 connection attempts.
+	err = attachProbe(e.module, "tcp_v6_connect", "trace_connect_entry")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = attachRetProbe(e.module, "tcp_v6_connect", "trace_connect_v6_return")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Open perf buffer and start processing IPv4 events.
+	e.v4EventCh, e.v4LostCh, err = openPerfBuffer(e.module, e.perfMaps, pageCount, "ipv4_events")
```

Comment:
Just curious if you considered having 1 buffer for all event types? Where you could have 2 tables:  one table for event payloads (byte stream), and another for event definitions containing an event type, payload size with an offset into the payload table. This way we can have a central message pump loop which would dispatch events for further processing. This part would be completely separated.

I think it would make it easier to add new events in the future where you would emit a new event in the bpf program and whoever processes the events may/may-not choose to handle it.

Basically, we can have:
1 way to define all event types
1 way to emit all types of events
1 way to pump all event types and dispatch them (where we can read in batches and dispatch to multiple handlers)